src/SlidingWindow.cpp

```diff
@@ -415,16 +579,112 @@ class SlidingWindow : public IRMutator {
         }
     }
 
+    Stmt visit(const For *op) override {
+        if (!(op->for_type == ForType::Serial || op->for_type == ForType::Unrolled)) {
+            return IRMutator::visit(op);
+        }
+        debug(3) << "Doing sliding window analysis on loop " << op->name << "\n";
+
+        string name = op->name;
+        Stmt body = op->body;
+        Expr loop_min = op->min;
+        Expr loop_extent = op->extent;
+        Expr loop_max = Variable::make(Int(32), op->name + ".loop_max");
+
+        Expr prev_loop_min = loop_min;
+        const Function *prev_func = nullptr;
+
+        list<pair<string, Expr>> new_lets;
+        for (const Function &func : sliding) {
+            debug(3) << "Doing sliding window analysis on function " << func.name() << "\n";
+
+            Expr sliding_loop_min;
+            if (prev_func && depends_on(func, *prev_func, body)) {
+                // The production of func depends on the production of prev_func.
+                // The loop min needs to grow to warm up func before prev_func.
+                sliding_loop_min = loop_min;
+            } else {
+                // The production of func does not depend on the production of prev_func.
+                // We can use the previous loop_min, and move the min to accommodate
+                // both func and prev_func.
+                sliding_loop_min = prev_loop_min;
+            }
+
+            SlidingWindowOnFunctionAndLoop slider(func, name, sliding_loop_min);
+            body = slider.mutate(body);
+
+            prev_loop_min = loop_min;
+            prev_func = &func;
+
+            if (slider.new_loop_min.defined()) {
+                // Update the loop body to use the adjusted loop min.
+                Expr new_loop_min = slider.new_loop_min;
+                if (!sliding_loop_min.same_as(loop_min)) {
+                    // If we didn't start from the loop min, take the union
+                    // of the new loop min and the loop min.
+                    new_loop_min = min(new_loop_min, loop_min);
+                }
+                string new_name = name + ".$n";
+                loop_min = Variable::make(Int(32), new_name + ".loop_min");
+                loop_extent = Variable::make(Int(32), new_name + ".loop_extent");
+                body = substitute({
```

Comment:
Here's where we'd have to rewrite prefetch directives. Should be a simple inline IRMutator. I think it's better than relying on an exact match to the string suffix above.